src/handlers/Firefox117.ts

```diff
@@ -0,0 +1,1093 @@
+import {
+	HandlerFactory,
+	HandlerInterface,
+	HandlerRunOptions,
+	HandlerSendOptions,
+	HandlerSendResult,
+	HandlerReceiveOptions,
+	HandlerReceiveResult,
+	HandlerSendDataChannelOptions,
+	HandlerSendDataChannelResult,
+	HandlerReceiveDataChannelOptions,
+	HandlerReceiveDataChannelResult
+} from './HandlerInterface';
+import { RemoteSdp } from './sdp/RemoteSdp';
+import { parse as parseScalabilityMode } from '../scalabilityModes';
+import { IceParameters, DtlsRole } from '../Transport';
+import { RtpCapabilities, RtpParameters } from '../RtpParameters';
+import { SctpCapabilities, SctpStreamParameters } from '../SctpParameters';
+
+const logger = new Logger('Firefox117');
+
+const SCTP_NUM_STREAMS = { OS: 16, MIS: 2048 };
+
+export class Firefox117 extends HandlerInterface
+{
+	// Handler direction.
+	private _direction?: 'send' | 'recv';
+	// Remote SDP handler.
+	private _remoteSdp?: RemoteSdp;
+	// Generic sending RTP parameters for audio and video.
+	private _sendingRtpParametersByKind?: { [key: string]: RtpParameters };
+	// Generic sending RTP parameters for audio and video suitable for the SDP
+	// remote answer.
+	private _sendingRemoteRtpParametersByKind?: { [key: string]: RtpParameters };
+	// RTCPeerConnection instance.
+	private _pc: any;
+	// Map of RTCTransceivers indexed by MID.
+	private readonly _mapMidTransceiver: Map<string, RTCRtpTransceiver> =
+		new Map();
+	// Local stream for sending.
+	private readonly _sendStream = new MediaStream();
+	// Whether a DataChannel m=application section has been created.
+	private _hasDataChannelMediaSection = false;
+	// Sending DataChannel id value counter. Incremented for each new DataChannel.
+	private _nextSendSctpStreamId = 0;
+	// Got transport local and remote parameters.
+	private _transportReady = false;
+
+	/**
+	 * Creates a factory function.
+	 */
+	static createFactory(): HandlerFactory
+	{
+		return (): Firefox117 => new Firefox117();
+	}
+
+	constructor()
+	{
+		super();
+	}
+
+	get name(): string
+	{
+		return 'Firefox117';
+	}
+
+	close(): void
+	{
+		logger.debug('close()');
+
+		// Close RTCPeerConnection.
+		if (this._pc)
+		{
+			try { this._pc.close(); }
+			catch (error) {}
+		}
+
+		this.emit('@close');
+	}
+
+	async getNativeRtpCapabilities(): Promise<RtpCapabilities>
+	{
+		logger.debug('getNativeRtpCapabilities()');
+
+		const pc = new (RTCPeerConnection as any)(
+			{
+				iceServers         : [],
+				iceTransportPolicy : 'all',
+				bundlePolicy       : 'max-bundle',
+				rtcpMuxPolicy      : 'require'
+			});
+
+		// NOTE: We need to add a real video track to get the RID extension mapping.
+		const canvas = document.createElement('canvas');
+
+		// NOTE: Otherwise Firefox fails in next line.
+		canvas.getContext('2d');
+
+		const fakeStream = (canvas as any).captureStream();
+		const fakeVideoTrack = fakeStream.getVideoTracks()[0];
+
+		try
+		{
+			pc.addTransceiver('audio', { direction: 'sendrecv' });
+
+			// START DEVIATION FROM FIREFOX 60
```

Comment:
ah nice, you found these comments that i left! i just now remembered that i left these as "breadcrumbs" to reminder myself (or any reader) how we got the changes introduced in https://github.com/daily-co/pluot-core/pull/9291.

Reply:
and https://github.com/daily-co/pluot-core/pull/9437

Reply:
Yes, I left the comments because the file is exactly as Firefox60 except where those comments are.